From cb7f4e93b78e8e623cc300784bfc543bc95ef611 Mon Sep 17 00:00:00 2001
From: Eugene Bolshakov <pub@relvarsoft.com>
Date: Sun, 6 Dec 2015 10:36:12 +0300
Subject: [PATCH 11/16] Quark: USB host

Linux Kernel 3.19 patch for Intel Galileo Gen1/Gen2 board
Source code is available on https://github.com/xbolshe/galileo-sources

Signed-off-by: Eugene Bolshakov <pub@relvarsoft.com>

Original source: 0015-Quark-USB-host-quark.patch
---
 drivers/usb/host/ehci-hub.c   |   78 +++++++++++++++++++++++++++++++++++-
 drivers/usb/host/ehci-pci.c   |   15 +++++++
 drivers/usb/host/ehci.h       |   16 ++++++++
 drivers/usb/host/pci-quirks.c |   88 +++++++++++++++++++++++++++++++++++++++++
 drivers/usb/host/pci-quirks.h |    4 ++
 5 files changed, 199 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/host/ehci-hub.c b/drivers/usb/host/ehci-hub.c
index 118edb7..773112e 100644
--- a/drivers/usb/host/ehci-hub.c
+++ b/drivers/usb/host/ehci-hub.c
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2001-2004 by David Brownell
+ * Portions Copyright (C) 2015 Intel Corporation
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the
@@ -588,6 +589,11 @@ static int check_reset_complete (
 	} else {
 		ehci_dbg(ehci, "port %d reset complete, port enabled\n",
 			index + 1);
+		if (ehci->has_x1000_phy && ehci->x1000_phy_squelch) {
+			/* If squelch was adjusted, then reset back to normal */
+			quirk_qrk_usb_phy_set_squelch(QRK_SQUELCH_DEFAULT);
+			ehci->x1000_phy_squelch = QRK_SQUELCH_DEFAULT;
+		}
 		/* ensure 440EPx ohci controller state is suspended */
 		if (ehci->has_amcc_usb23)
 			set_ohci_hcfs(ehci, 0);
@@ -680,6 +686,62 @@ ehci_hub_status_data (struct usb_hcd *hcd, char *buf)
 
 /*-------------------------------------------------------------------------*/
 
+/*
+ * __ehci_adjust_x1000_squelch
+ *
+ * Adjust the squelch on PHY for Intel Quark X1000
+ *
+ * Feature on X1000 SOC where squelch of on SOC PHY may need to be adjusted
+ * during enumeration of HS, trying a number of different levels.
+ * The adjustment is followed by a port repower to enable
+ * the state machine realign and a RESET to restart the
+ * enumeration process. This will result in succesful HS enumeration instead
+ * of cycling through waits and resets which were not solving the issue and
+ * ultimately ending up with no enumeration.
+ */
+
+static void
+__ehci_adjust_x1000_squelch(
+	struct ehci_hcd	*ehci,
+	u32 __iomem	*status_reg,
+	u16		windex
+) {
+	u32		temp;
+	temp = ehci_readl(ehci, status_reg);
+
+	switch (ehci->x1000_phy_squelch) {
+	case QRK_SQUELCH_DEFAULT:   /* No squelch yet done now put it low */
+		ehci->x1000_phy_squelch = QRK_SQUELCH_LO;
+		break;
+	case QRK_SQUELCH_LO: /* Low squelch done now put it high */
+		ehci->x1000_phy_squelch = QRK_SQUELCH_HI;
+		break;
+	case QRK_SQUELCH_HI:
+			/* High done now put it back to default and
+			   reset state for for next time */
+		ehci->x1000_phy_squelch = QRK_SQUELCH_DEFAULT;
+		break;
+	}
+	temp = ehci_readl(ehci, status_reg);
+	ehci_writel(ehci, temp & ~PORT_POWER, status_reg);
+	quirk_qrk_usb_phy_set_squelch(ehci->x1000_phy_squelch);
+	temp = ehci_readl(ehci, status_reg);
+	ehci_writel(ehci, temp | PORT_POWER, status_reg);
+	ehci_handshake(ehci, status_reg, PORT_CONNECT, 1, 2000);
+			/*  allow time for repower to work */
+	if (ehci->x1000_phy_squelch == QRK_SQUELCH_DEFAULT)
+		return;	/* do not reset port when finished
+			   as we are OK with enum. Just set
+			   back to default for next device */
+	temp = ehci_readl(ehci, status_reg);
+	temp |= PORT_RESET;
+	temp &= ~PORT_PE;
+	ehci->reset_done[windex] = jiffies + msecs_to_jiffies(50);
+	ehci_writel(ehci, temp, status_reg);
+}
+
+/*-------------------------------------------------------------------------*/
+
 static void
 ehci_hub_descriptor (
 	struct ehci_hcd			*ehci,
@@ -1063,7 +1125,16 @@ int ehci_hub_control(
 			 */
 			retval = ehci_handshake(ehci, status_reg,
 					PORT_RESET, 0, 1000);
-			if (retval != 0) {
+			if (retval == -ETIMEDOUT && ehci->has_x1000_phy) {
+				/* Quark, squelch adjust and RESET needed */
+				__ehci_adjust_x1000_squelch(ehci, status_reg,
+								wIndex);
+				if (ehci->x1000_phy_squelch)
+					/* If squelch adjusted */
+					goto error;
+				/* else squelch is back at default allow enum
+					to continue */
+			} else if (retval != 0) {
 				ehci_err (ehci, "port %d reset error %d\n",
 					wIndex + 1, retval);
 				goto error;
@@ -1210,7 +1281,10 @@ int ehci_hub_control(
 			} else {
 				temp |= PORT_RESET;
 				temp &= ~PORT_PE;
-
+				if (ehci->has_x1000_phy
+						&& ehci->x1000_phy_squelch)
+					ehci->x1000_phy_squelch
+						= QRK_SQUELCH_DEFAULT;
 				/*
 				 * caller must wait, then call GetPortStatus
 				 * usb 2.0 spec says 50 ms resets on root
diff --git a/drivers/usb/host/ehci-pci.c b/drivers/usb/host/ehci-pci.c
index 851006a..a584f5e 100644
--- a/drivers/usb/host/ehci-pci.c
+++ b/drivers/usb/host/ehci-pci.c
@@ -75,6 +75,10 @@ static int ehci_pci_reinit(struct ehci_hcd *ehci, struct pci_dev *pdev)
 			ehci->regs->intel_quark_x1000_insnreg01);
 	}
 
+	/* Reset the threshold limit */
+	if(unlikely(usb_is_intel_qrk(pdev)))
+		usb_set_qrk_bulk_thresh(pdev);
+
 	return 0;
 }
 
@@ -137,6 +141,17 @@ static int ehci_pci_setup(struct usb_hcd *hcd)
 	case PCI_VENDOR_ID_INTEL:
 		if (pdev->device == PCI_DEVICE_ID_INTEL_CE4100_USB)
 			hcd->has_tt = 1;
+		else if (pdev->device == PCI_DEVICE_ID_INTEL_QUARK_X1000_SOC
+						&& pdev->revision  == 0x10) {
+			ehci->has_x1000_phy = 1;
+			ehci->x1000_phy_squelch = QRK_SQUELCH_DEFAULT;
+			/* Intel Quark device of this revision requires
+				adjustment of squelch on the PHY
+			*/
+			ehci_info(ehci,
+			"QUARK USB phy rev 0x%x may have squelch ref adjusted\n"
+						, pdev->revision);
+		}
 		break;
 	case PCI_VENDOR_ID_TDI:
 		if (pdev->device == PCI_DEVICE_ID_TDI_EHCI)
diff --git a/drivers/usb/host/ehci.h b/drivers/usb/host/ehci.h
index 6f0577b..86d8b6f 100644
--- a/drivers/usb/host/ehci.h
+++ b/drivers/usb/host/ehci.h
@@ -227,6 +227,22 @@ struct ehci_hcd {			/* one per controller */
 	unsigned		need_oc_pp_cycle:1; /* MPC834X port power */
 	unsigned		imx28_write_fix:1; /* For Freescale i.MX28 */
 
+	unsigned		has_x1000_phy:1;
+	/* Intel Quark X10xx needs squelch adjustment during HS enumeration
+	 * to eliminate noise at around default power on squelch of 112.5mV
+	 * which causes the noise to look like signalling. We need to adjust
+	 * the squelch to eliminate that false signalling.
+	 * First we adjust it lo and if the false signalling is now damped,
+	 * enumeration works OK. If that fails  we adjust hi and see
+	 * if it enumerates OK, if not we are at the END and will default to
+	 * full speed.  */
+	#define QRK_SQUELCH_DEFAULT	0 /* apply default of 112.5 mV */
+	#define QRK_SQUELCH_LO		1 /* apply low  of 100 mV */
+	#define QRK_SQUELCH_HI		2 /* apply high of 125 mV */
+
+	unsigned		x1000_phy_squelch:2;
+					/*Squelch state during quirk */
+
 	/* required for usb32 quirk */
 	#define OHCI_CTRL_HCFS          (3 << 6)
 	#define OHCI_USB_OPER           (2 << 6)
diff --git a/drivers/usb/host/pci-quirks.c b/drivers/usb/host/pci-quirks.c
index dd483c1..bd9a0de 100644
--- a/drivers/usb/host/pci-quirks.c
+++ b/drivers/usb/host/pci-quirks.c
@@ -16,6 +16,7 @@
 #include <linux/export.h>
 #include <linux/acpi.h>
 #include <linux/dmi.h>
+#include <linux/intel_qrk_sb.h>
 #include "pci-quirks.h"
 #include "xhci-ext-caps.h"
 
@@ -832,6 +833,52 @@ static int handshake(void __iomem *ptr, u32 mask, u32 done,
 	return -ETIMEDOUT;
 }
 
+bool usb_is_intel_qrk(struct pci_dev *pdev)
+{
+	return pdev->vendor == PCI_VENDOR_ID_INTEL &&
+		pdev->device == PCI_DEVICE_ID_INTEL_QUARK_X1000_SOC;
+
+}
+EXPORT_SYMBOL_GPL(usb_is_intel_qrk);
+
+#define EHCI_INSNREG01			0x84
+#define EHCI_INSNREG01_THRESH	0x007F007F	/* Threshold value */
+void usb_set_qrk_bulk_thresh(struct pci_dev *pdev)
+{
+	void __iomem *base, *op_reg_base;
+	u8 cap_length;
+	u32 val;
+
+	if (!mmio_resource_enabled(pdev, 0))
+		return;
+
+	base = pci_ioremap_bar(pdev, 0);
+	if (base == NULL)
+		return;
+
+	cap_length = readb(base);
+	op_reg_base = base + cap_length;
+
+	val = readl(op_reg_base + EHCI_INSNREG01);
+	dev_printk(KERN_INFO, &pdev->dev, "INSNREG01 is 0x%08x\n", val);
+
+	val = EHCI_INSNREG01_THRESH;
+
+	writel(val, op_reg_base + EHCI_INSNREG01);
+
+	val = readl(op_reg_base + EHCI_INSNREG01);
+	dev_printk(KERN_INFO, &pdev->dev, "INSNREG01 is 0x%08x\n", val);
+
+	iounmap(base);
+
+}
+EXPORT_SYMBOL_GPL(usb_set_qrk_bulk_thresh);
+
+
+
+
+
+
 /*
  * Intel's Panther Point chipset has two host controllers (EHCI and xHCI) that
  * share some number of ports.  These ports can be switched between either
@@ -1048,6 +1095,47 @@ hc_init:
 
 	iounmap(base);
 }
+#ifdef CONFIG_INTEL_QUARK_X1000_SOC
+/**
+ * quirk_qrk_usb_phy_set_squelch
+ *
+ * Uses side band access on quark to access USB PHY registers where the
+ * squelch value can be adjusted.
+ * @threshold: ref to millivolts to set the squelch to there are just a few
+ *             values available to use in quark
+ *
+ * QRK_SQUELCH_DEFAULT	0 apply default of 112.5 mV
+ * QRK_SQUELCH_LO	1 apply low  of 100 mV
+ * QRK_SQUELCH_HI	2 apply high of 125 mV
+ */
+#define USB2COMPBG	0x7F04	/* PHY register over side band */
+#define HS_SQ_REF_POS   13	/* bit position for squelch */
+#define HS_SQ_REF_MASK  (3 << HS_SQ_REF_POS) /* bit mask for squelch */
+
+#define USBPHY_SB_READ	0x06	/* Sideband read command  */
+#define USBPHY_SB_WRITE	0x07	/* Sideband write command */
+#define SB_ID_USBPHY	0x14	/* Port of USB PHY */
+
+void quirk_qrk_usb_phy_set_squelch(u32 threshold)
+{
+	u32 regval, regnew;
+	intel_qrk_sb_read_reg(SB_ID_USBPHY, USBPHY_SB_READ, USB2COMPBG,
+								 &regval, 1);
+	regnew = regval & ~HS_SQ_REF_MASK;
+	regnew |= ((threshold<<HS_SQ_REF_POS) & HS_SQ_REF_MASK);
+	intel_qrk_sb_write_reg(SB_ID_USBPHY, USBPHY_SB_WRITE, USB2COMPBG,
+								regnew, 1);
+	pr_info("USB PHY squelch ref adjusted from %8x to %8x\n",
+						regval, regnew);
+}
+#else
+inline void quirk_qrk_usb_phy_set_squelch(u32 threshold)
+{
+}
+#endif
+EXPORT_SYMBOL_GPL(quirk_qrk_usb_phy_set_squelch);
+
+
 
 static void quirk_usb_early_handoff(struct pci_dev *pdev)
 {
diff --git a/drivers/usb/host/pci-quirks.h b/drivers/usb/host/pci-quirks.h
index c622ddf..b95eee1 100644
--- a/drivers/usb/host/pci-quirks.h
+++ b/drivers/usb/host/pci-quirks.h
@@ -11,8 +11,12 @@ bool usb_amd_prefetch_quirk(void);
 void usb_amd_dev_put(void);
 void usb_amd_quirk_pll_disable(void);
 void usb_amd_quirk_pll_enable(void);
+bool usb_is_intel_qrk(struct pci_dev *pdev);
+void usb_set_qrk_bulk_thresh(struct pci_dev *pdev);
 void usb_enable_intel_xhci_ports(struct pci_dev *xhci_pdev);
 void usb_disable_xhci_ports(struct pci_dev *xhci_pdev);
+#define PCI_DEVICE_ID_INTEL_QUARK_X1000_SOC		0x0939
+void quirk_qrk_usb_phy_set_squelch(u32 threshold);
 void sb800_prefetch(struct device *dev, int on);
 #else
 struct pci_dev;
-- 
1.7.9.5

